osf_tests/test_comment.py

```diff
@@ -979,7 +979,7 @@ def test_comments_move_when_file_moved_from_project_to_component(self, project,
         self.file.move_under(destination['node'].get_addon(self.provider).get_root())
         payload = self._create_payload('move', user, source, destination, self.file._id)
         update_file_guid_referent(self=None, target=destination['node'], event_type='addon_file_moved', payload=payload)
-        self.guid.reload()
+        self.guid.referent.reload()
```

Comment:
This fix makes sense since Django `refresh_from_db()` (called by `reload()`) doesn't update relationship field and in this case a GFK.

My concern is, given that we use `refresh_from_db()` or `reload()` a lot in the actual code, this behavior change due to the upgrade may led to similar problems not covered by the tests.

Behavior of 1.11: [doc](https://docs.djangoproject.com/en/1.11/ref/models/instances/#django.db.models.Model.refresh_from_db), [code](https://docs.djangoproject.com/en/1.11/_modules/django/db/models/base/#Model.refresh_from_db)
Behavior of 3.2] [doc](https://docs.djangoproject.com/en/3.2/_modules/django/db/models/base/#Model.refresh_from_db), [code](https://docs.djangoproject.com/en/3.2/ref/models/instances/#django.db.models.Model.refresh_from_db)

In addition, [our customization of the 1.11 `refresh_from_db()`](https://github.com/CenterForOpenScience/osf.io/blob/d07206d3a7b4a26e440c66131531627170e9b1c9/osf/models/base.py#L110-L115) clears cache only, which seems to have been added in [2.2](https://docs.djangoproject.com/en/2.2/ref/models/instances/#django.db.models.Model.refresh_from_db)

> Any cached relations are cleared from the reloaded instance.

Reply:
More Update:

[We were using `private_fields` as GFK](https://github.com/CenterForOpenScience/osf.io/blob/d07206d3a7b4a26e440c66131531627170e9b1c9/osf/models/base.py#L113) which turns out to be wrong. Although GFKs are returned in `private_fields` but the correct way to check GFK should be

```python
f.is_relation and f.many_to_one and not (hasattr(f.remote_field, 'model') and f.remote_field.model)
```

which, can be found in private APIs from the source code of [django.db.models.options](https://docs.djangoproject.com/en/3.2/_modules/django/db/models/options/).

And `refresh_from_db()` [only refreshes non-deferred `concrete_fields`](https://docs.djangoproject.com/en/3.2/_modules/django/db/models/base/) when no `field` are provided.

An issue with refreshing related fields is that it can cause infinite loop, as confirmed in local shell.
Another issue is when relation is many-to-many or one-to-many, things become more complicated.
This is probably why `django` doesn't do it but recommend reload the relate field(s) explicitly when necessary.
The best we can do is to refresh only one-to-one or many-to-ones.
As for this case, let's reload GFK's only.